<img src='https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQ-VfNtOyJbsaxu43Kztf_cv1mgBG6ZIQZEVw&usqp=CAU'>

# Procesamiento de Lenguage Natural

## Taller #5: TF-IDF

## Presentado por: Juan Sebastian Cruz Dionicio
`Fecha de entrega: Septiembre 19, 2020. (Antes del inicio de la próxima clase).`

`Modo de entrega: Subir link de GitHub al aula virtual.`

In [2]:
import re
import numpy as np
import pandas as pd

from nltk.corpus import stopwords
stopwords_sp = stopwords.words('spanish')

from nltk.stem.snowball import SnowballStemmer
spanishStemmer=SnowballStemmer("spanish")

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Punto 1: Pre-Procesamiento

- `[18 pts]` Leer el archivo `Princesas.csv` usando `pandas` y crear una nueva columna con el texto en minúscula, sin caracteres especiales ni números, sin palabras vacias y hacer stemming de las palabras

In [3]:
import os
os.getcwd()

df = pd.read_csv("princesas.csv", sep=",")

df.head(10)

Princesa                                       Personalidad
0  Blancanieves  Blancanieves es una princesa de noble cuna que...
1    Cenicienta  Cenicienta es inicialmente una sirvienta en su...
2        Aurora  La Princesa Aurora, la Bella Durmiente, es la ...
3         Bella  Bella es una muchacha que vive en la campiña f...
4        Jasmín  Cuando se introdujo por primera vez, la Prince...
5    Pocahontas  El nombre de Pocahontas significa "Pequeña Sil...
6         Mulan  Mulan es atípica a los anteriores papeles feme...
7         Tiana  Es una joven camarera que sueña con ser dueña ...
8        Mérida  Mérida llama la atención por su característico...
9         Moana  Moana, una joven de 16 años de edad, hija únic...

In [33]:
# data

Princesa                                       Personalidad
0  Blancanieves  Blancanieves es una princesa de noble cuna que...
1    Cenicienta  Cenicienta es inicialmente una sirvienta en su...
2        Aurora  La Princesa Aurora, la Bella Durmiente, es la ...
3         Bella  Bella es una muchacha que vive en la campiña f...
4        Jasmín  Cuando se introdujo por primera vez, la Prince...
5    Pocahontas  El nombre de Pocahontas significa "Pequeña Sil...
6         Mulan  Mulan es atípica a los anteriores papeles feme...
7         Tiana  Es una joven camarera que sueña con ser dueña ...
8        Mérida  Mérida llama la atención por su característico...
9         Moana  Moana, una joven de 16 años de edad, hija únic...

In [64]:
def pre_procesado(texto):
    texto = texto.lower()
    texto = re.sub(r"[\W\d_]+", " ", texto)
    texto = " ".join([palabra for palabra in texto.split() if palabra not in stopwords_sp])
    return texto

In [97]:
df['pre-procesado'] = df['Personalidad'].apply(lambda texto: pre_procesado(texto))
df

Princesa                                       Personalidad  \
0  Blancanieves  Blancanieves es una princesa de noble cuna que...   
1    Cenicienta  Cenicienta es inicialmente una sirvienta en su...   
2        Aurora  La Princesa Aurora, la Bella Durmiente, es la ...   
3         Bella  Bella es una muchacha que vive en la campiña f...   
4        Jasmín  Cuando se introdujo por primera vez, la Prince...   
5    Pocahontas  El nombre de Pocahontas significa "Pequeña Sil...   
6         Mulan  Mulan es atípica a los anteriores papeles feme...   
7         Tiana  Es una joven camarera que sueña con ser dueña ...   
8        Mérida  Mérida llama la atención por su característico...   
9         Moana  Moana, una joven de 16 años de edad, hija únic...   

                                       pre-procesado  
0  blancanieves princesa noble cuna ve forzada se...  
1  cenicienta inicialmente sirvienta casa constan...  
2  princesa aurora bella durmiente hija única rei...  
3  bella muchacha vive campiña francesa padre inv...  
4  introdujo primera vez princesa jasmín pocos dí...  
5  nombre pocahontas significa pequeña silenciosa...  
6  mulan atípica anteriores papeles femeninos pel...  
7  joven camarera sueña ser dueña propio restaura...  
8  mérida llama atención característico pelo anar...  
9  moana joven años edad hija única sucesora impo...

In [98]:
df['pre-procesado'] = df['pre-procesado'].str.split()
df['pre-procesado'] = df['pre-procesado'].apply(lambda x: ' '.join([spanishStemmer.stem(y) for y in x]))
df

Princesa                                       Personalidad  \
0  Blancanieves  Blancanieves es una princesa de noble cuna que...   
1    Cenicienta  Cenicienta es inicialmente una sirvienta en su...   
2        Aurora  La Princesa Aurora, la Bella Durmiente, es la ...   
3         Bella  Bella es una muchacha que vive en la campiña f...   
4        Jasmín  Cuando se introdujo por primera vez, la Prince...   
5    Pocahontas  El nombre de Pocahontas significa "Pequeña Sil...   
6         Mulan  Mulan es atípica a los anteriores papeles feme...   
7         Tiana  Es una joven camarera que sueña con ser dueña ...   
8        Mérida  Mérida llama la atención por su característico...   
9         Moana  Moana, una joven de 16 años de edad, hija únic...   

                                       pre-procesado  
0  blancaniev princes nobl cun ve forz servidumbr...  
1  cenicient inicial sirvient cas constant objet ...  
2  princes auror bell durmient hij unic rein flor...  
3  bell muchach viv campiñ frances padr inventor ...  
4  introduj primer vez princes jasmin poc dias de...  
5  nombr pocahont signif pequeñ silenci bas figur...  
6  mul atip anterior papel femenin pelicul disney...  
7  jov camarer sueñ ser dueñ propi restaur algun ...  
8  mer llam atencion caracterist pel anaranj oscu...  
9  moan jov años edad hij unic sucesor import jef...

In [36]:
# data procesada

Princesa                                       Personalidad  \
0  Blancanieves  Blancanieves es una princesa de noble cuna que...   
1    Cenicienta  Cenicienta es inicialmente una sirvienta en su...   
2        Aurora  La Princesa Aurora, la Bella Durmiente, es la ...   
3         Bella  Bella es una muchacha que vive en la campiña f...   
4        Jasmín  Cuando se introdujo por primera vez, la Prince...   
5    Pocahontas  El nombre de Pocahontas significa "Pequeña Sil...   
6         Mulan  Mulan es atípica a los anteriores papeles feme...   
7         Tiana  Es una joven camarera que sueña con ser dueña ...   
8        Mérida  Mérida llama la atención por su característico...   
9         Moana  Moana, una joven de 16 años de edad, hija únic...   

                                       pre-procesado  
0  blancaniev princes nobl cun ve forz servidumbr...  
1  cenicient inicial sirvient cas constant objet ...  
2  princes auror bell durmient hij unic rein flor...  
3  bell muchach viv campiñ frances padr inventor ...  
4  introduj primer vez princes jasmin poc dias de...  
5  nombr pocahont signif pequeñ silenci bas figur...  
6  mul atip anterior papel femenin pelicul disney...  
7  jov camarer sueñ ser dueñ propi restaur algun ...  
8  mer llam atencion caracterist pel anaranj oscu...  
9  moan jov años edad hij unic sucesor import jef...

# Punto 2: TF-IDF

- `[16 pts]` Crear la matriz TF-IDF

In [113]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer()
tfidf = tfidf_vect.fit_transform(df['pre-procesado'].values)

tfidf_matrix = pd.DataFrame(tfidf.toarray(), columns=tfidf_vect.get_feature_names())
tfidf_matrix.index=df['Princesa'].values

tfidf_matrix=tfidf_matrix.T.round(3)

tfidf_matrix

Blancanieves  Cenicienta  Aurora  Bella  Jasmín  Pocahontas  Mulan  \
abrum           0.000         0.0   0.065  0.000   0.000       0.000    0.0   
abuel           0.000         0.0   0.000  0.000   0.000       0.000    0.0   
acuerd          0.000         0.0   0.056  0.000   0.000       0.000    0.0   
ademas          0.000         0.0   0.000  0.000   0.000       0.099    0.0   
afortun         0.087         0.0   0.000  0.000   0.000       0.000    0.0   
...               ...         ...     ...    ...     ...         ...    ...   
viv             0.000         0.0   0.000  0.074   0.000       0.000    0.0   
volunt          0.000         0.0   0.000  0.000   0.161       0.000    0.0   
volv            0.000         0.0   0.065  0.000   0.000       0.000    0.0   
vudu            0.000         0.0   0.000  0.000   0.000       0.000    0.0   
wak             0.000         0.0   0.000  0.000   0.000       0.000    0.0   

         Tiana  Mérida  Moana  
abrum    0.000    0.00  0.000  
abuel    0.000    0.00  0.102  
acuerd   0.000    0.08  0.000  
ademas   0.000    0.00  0.000  
afortun  0.000    0.00  0.000  
...        ...     ...    ...  
viv      0.000    0.00  0.087  
volunt   0.000    0.00  0.000  
volv     0.000    0.00  0.000  
vudu     0.122    0.00  0.000  
wak      0.000    0.00  0.102  

[569 rows x 10 columns]

In [44]:
# tfidf_matrix

Blancanieves  Cenicienta  Aurora  Bella  Jasmín  Pocahontas  Mulan  \
abrum           0.000         0.0   0.065  0.000   0.000       0.000    0.0   
abuel           0.000         0.0   0.000  0.000   0.000       0.000    0.0   
acuerd          0.000         0.0   0.056  0.000   0.000       0.000    0.0   
ademas          0.000         0.0   0.000  0.000   0.000       0.099    0.0   
afortun         0.087         0.0   0.000  0.000   0.000       0.000    0.0   
...               ...         ...     ...    ...     ...         ...    ...   
viv             0.000         0.0   0.000  0.074   0.000       0.000    0.0   
volunt          0.000         0.0   0.000  0.000   0.161       0.000    0.0   
volv            0.000         0.0   0.065  0.000   0.000       0.000    0.0   
vudu            0.000         0.0   0.000  0.000   0.000       0.000    0.0   
wak             0.000         0.0   0.000  0.000   0.000       0.000    0.0   

         Tiana  Mérida  Moana  
abrum    0.000    0.00  0.000  
abuel    0.000    0.00  0.102  
acuerd   0.000    0.08  0.000  
ademas   0.000    0.00  0.000  
afortun  0.000    0.00  0.000  
...        ...     ...    ...  
viv      0.000    0.00  0.087  
volunt   0.000    0.00  0.000  
volv     0.000    0.00  0.000  
vudu     0.122    0.00  0.000  
wak      0.000    0.00  0.102  

[569 rows x 10 columns]

# Punto 3: Distancia del coseno
- `[12 pts]` Calcular la distancia del coseno entre cada una de las princesas
- `[2 pts]` ¿Cuáles son las princesas más parecidas?
- `[2 pts]` ¿Cuáles son las princesas más diferentes?

In [114]:
from sklearn.metrics.pairwise import cosine_distances
 
dist_cos = cosine_distances(tfidf_matrix.T.values)

dist_cos = pd.DataFrame(dist_cos, columns = tfidf_matrix.columns, index = tfidf_matrix.columns)

dist_cos

Blancanieves  Cenicienta    Aurora     Bella    Jasmín  \
Blancanieves      0.000000    0.846420  0.842223  0.918112  0.935103   
Cenicienta        0.846420    0.000000  0.858412  0.947243  0.959209   
Aurora            0.842223    0.858412  0.000000  0.832503  0.937195   
Bella             0.918112    0.947243  0.832503  0.000000  0.888219   
Jasmín            0.935103    0.959209  0.937195  0.888219  0.000000   
Pocahontas        0.922165    0.960479  0.868994  0.890513  0.942753   
Mulan             0.884147    0.932663  0.919500  0.882169  0.935198   
Tiana             0.960082    0.918639  0.939231  0.973338  0.937868   
Mérida            0.985136    0.970839  0.940723  0.943607  0.972505   
Moana             0.951298    0.974925  0.965930  0.923546  0.968929   

              Pocahontas     Mulan     Tiana    Mérida     Moana  
Blancanieves    0.922165  0.884147  0.960082  0.985136  0.951298  
Cenicienta      0.960479  0.932663  0.918639  0.970839  0.974925  
Aurora          0.868994  0.919500  0.939231  0.940723  0.965930  
Bella           0.890513  0.882169  0.973338  0.943607  0.923546  
Jasmín          0.942753  0.935198  0.937868  0.972505  0.968929  
Pocahontas      0.000000  0.888844  0.968918  0.979646  0.936091  
Mulan           0.888844  0.000000  0.991170  0.987267  0.945844  
Tiana           0.968918  0.991170  0.000000  0.979275  0.971443  
Mérida          0.979646  0.987267  0.979275  0.000000  0.971818  
Moana           0.936091  0.945844  0.971443  0.971818  0.000000

In [47]:
# dist_cos

Blancanieves  Cenicienta    Aurora     Bella    Jasmín  \
Blancanieves      0.000000    0.846420  0.842223  0.918112  0.935103   
Cenicienta        0.846420    0.000000  0.858412  0.947243  0.959209   
Aurora            0.842223    0.858412  0.000000  0.832503  0.937195   
Bella             0.918112    0.947243  0.832503  0.000000  0.888219   
Jasmín            0.935103    0.959209  0.937195  0.888219  0.000000   
Pocahontas        0.922165    0.960479  0.868994  0.890513  0.942753   
Mulan             0.884147    0.932663  0.919500  0.882169  0.935198   
Tiana             0.960082    0.918639  0.939231  0.973338  0.937868   
Mérida            0.985136    0.970839  0.940723  0.943607  0.972505   
Moana             0.951298    0.974925  0.965930  0.923546  0.968929   

              Pocahontas     Mulan     Tiana    Mérida     Moana  
Blancanieves    0.922165  0.884147  0.960082  0.985136  0.951298  
Cenicienta      0.960479  0.932663  0.918639  0.970839  0.974925  
Aurora          0.868994  0.919500  0.939231  0.940723  0.965930  
Bella           0.890513  0.882169  0.973338  0.943607  0.923546  
Jasmín          0.942753  0.935198  0.937868  0.972505  0.968929  
Pocahontas     -0.000000  0.888844  0.968918  0.979646  0.936091  
Mulan           0.888844  0.000000  0.991170  0.987267  0.945844  
Tiana           0.968918  0.991170  0.000000  0.979275  0.971443  
Mérida          0.979646  0.987267  0.979275  0.000000  0.971818  
Moana           0.936091  0.945844  0.971443  0.971818  0.000000

### - ¿Cuáles son las princesas más parecidas?
Las dos princesas mas parecidas son Mulan y Tiana con un porcentaje del 99.12%, sin embargo en general las ultimas tres princesas de la tabla (Tiana, Merida y Moana) se parecen mucho con todas las demas. 

### -¿Cuáles son las princesas más diferentes?
Las princesas con el porcentaje mas bajo y por lo tanto mas diferentes son Bella y Aurora con un porcentaje del 83.2%, seguidas por Blancanieves y Cenicienta con un porcentaje de 84.6%. 